File: node/scripts/test_weather_tool.ts
#!/usr/bin/env ts-node
/**
 * Test script for weather tool functionality.
 * Tests that providers correctly call the weather API with lat/lon/location_name.
 *
 * Usage:
 *     ts-node test_weather_tool.ts                      # Run all tests
 *     ts-node test_weather_tool.ts --provider anthropic # Test specific provider
 *     ts-node test_weather_tool.ts --list               # List available providers
 */

import * as dotenv from 'dotenv';
import * as path from 'path';
import { fileURLToPath } from 'url';
import { PostHog } from 'posthog-node';
import { AnthropicProvider } from '../dist/providers/anthropic.js';
import { AnthropicStreamingProvider } from '../dist/providers/anthropic-streaming.js';
import { GeminiProvider } from '../dist/providers/gemini.js';
import { GeminiStreamingProvider } from '../dist/providers/gemini-streaming.js';
import { LangChainProvider } from '../dist/providers/langchain.js';
import { OpenAIProvider } from '../dist/providers/openai.js';
import { OpenAIChatProvider } from '../dist/providers/openai-chat.js';
import { OpenAIChatStreamingProvider } from '../dist/providers/openai-chat-streaming.js';
import { OpenAIStreamingProvider } from '../dist/providers/openai-streaming.js';
import { VercelAIProvider } from '../dist/providers/vercel-ai.js';
import { VercelAIStreamingProvider } from '../dist/providers/vercel-ai-streaming.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Load environment variables
dotenv.config({ path: path.join(__dirname, '..', '..', '.env') });

// Define available providers
interface ProviderConfig {
  class: any;
  name: string;
  isStreaming: boolean;
  requiresThinking?: boolean;
}

const AVAILABLE_PROVIDERS: Record<string, ProviderConfig> = {
  'anthropic': {
    class: AnthropicProvider,
    name: 'Anthropic Provider',
    isStreaming: false,
    requiresThinking: true
  },
  'anthropic-streaming': {
    class: AnthropicStreamingProvider,
    name: 'Anthropic Streaming Provider',
    isStreaming: true,
    requiresThinking: true
  },
  'gemini': {
    class: GeminiProvider,
    name: 'Google Gemini Provider',
    isStreaming: false
  },
  'gemini-streaming': {
    class: GeminiStreamingProvider,
    name: 'Google Gemini Streaming Provider',
    isStreaming: true
  },
  'openai': {
    class: OpenAIProvider,
    name: 'OpenAI Responses API Provider',
    isStreaming: false
  },
  'openai-streaming': {
    class: OpenAIStreamingProvider,
    name: 'OpenAI Responses API Streaming Provider',
    isStreaming: true
  },
  'openai-chat': {
    class: OpenAIChatProvider,
    name: 'OpenAI Chat Completions Provider',
    isStreaming: false
  },
  'openai-chat-streaming': {
    class: OpenAIChatStreamingProvider,
    name: 'OpenAI Chat Completions Streaming Provider',
    isStreaming: true
  },
  'langchain': {
    class: LangChainProvider,
    name: 'LangChain Provider',
    isStreaming: false
  },
  'vercel-ai': {
    class: VercelAIProvider,
    name: 'Vercel AI SDK Provider',
    isStreaming: false
  },
  'vercel-ai-streaming': {
    class: VercelAIStreamingProvider,
    name: 'Vercel AI SDK Streaming Provider',
    isStreaming: true
  },
};

function initPostHog(): PostHog | null {
  const posthogApiKey = process.env.POSTHOG_API_KEY;
  const posthogHost = process.env.POSTHOG_HOST || 'https://us.i.posthog.com';

  if (!posthogApiKey) {
    console.log('⚠️  Warning: POSTHOG_API_KEY not set, analytics will not be tracked');
    return null;
  }

  return new PostHog(posthogApiKey, {
    host: posthogHost,
    flushAt: 1,
    flushInterval: 0
  });
}

async function testProvider(
  providerClass: any,
  providerName: string,
  useStreaming: boolean,
  requiresThinking: boolean = false
): Promise<boolean> {
  console.log(`\n${'='.repeat(80)}`);
  console.log(`🧪 Testing ${providerName}`);
  console.log(`${'='.repeat(80)}\n`);

  // Initialize PostHog
  const posthogClient = initPostHog();

  if (!posthogClient) {
    console.log('❌ Cannot run test without PostHog client');
    return false;
  }

  // Initialize provider (handle Anthropic's different constructor signature)
  const provider = requiresThinking
    ? new providerClass(posthogClient, false, undefined, null)
    : new providerClass(posthogClient, null);

  // Test query
  const testQuery = "What's the weather in Dublin, Ireland?";
  console.log(`📝 Query: ${testQuery}\n`);

  try {
    let response: string;

    if (useStreaming) {
      console.log('🔄 Streaming response:\n');
      let fullResponse = '';
      const stream = provider.chatStream(testQuery);

      for await (const chunk of stream) {
        process.stdout.write(chunk);
        fullResponse += chunk;
      }
      console.log('\n');
      response = fullResponse;
    } else {
      console.log('💬 Response:\n');
      response = await provider.chat(testQuery);
      console.log(response);
      console.log();
    }

    // Check if response contains weather data
    if (response.includes('°C') || response.includes('°F')) {
      console.log('✅ Test PASSED - Weather data received!');

      // Check if location name is used
      if (response.includes('Dublin')) {
        console.log('✅ Location name displayed correctly!');
      } else {
        console.log('⚠️  Warning: Location name might not be displayed');
      }

      return true;
    } else {
      console.log('❌ Test FAILED - No weather data in response');
      return false;
    }

  } catch (error: any) {
    console.log(`❌ Test FAILED with error: ${error.message}`);
    console.error(error.stack);
    return false;
  } finally {
    // Shutdown PostHog
    if (posthogClient) {
      await posthogClient.shutdown();
    }
  }
}

function listProviders(): void {
  console.log(`\n${'='.repeat(80)}`);
  console.log('📋 Available Providers');
  console.log(`${'='.repeat(80)}\n`);

  for (const [key, config] of Object.entries(AVAILABLE_PROVIDERS)) {
    const streamIndicator = config.isStreaming ? ' (streaming)' : '';
    console.log(`  • ${key.padEnd(25)} - ${config.name}${streamIndicator}`);
  }

  console.log();
}

async function main() {
  const args = process.argv.slice(2);

  // Parse arguments
  let specificProvider: string | null = null;
  let showList = false;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--provider' || args[i] === '-p') {
      specificProvider = args[i + 1];
      i++;
    } else if (args[i] === '--list' || args[i] === '-l') {
      showList = true;
    }
  }

  // Handle --list flag
  if (showList) {
    listProviders();
    process.exit(0);
  }

  // Validate provider if specified
  if (specificProvider && !AVAILABLE_PROVIDERS[specificProvider]) {
    console.error(`❌ Unknown provider: ${specificProvider}`);
    console.log('\nAvailable providers:');
    listProviders();
    process.exit(1);
  }

  console.log(`\n${'='.repeat(80)}`);
  console.log('🚀 Weather Tool Test Suite');
  console.log(`${'='.repeat(80)}`);

  const results: Array<[string, boolean]> = [];

  // Determine which providers to test
  const providersToTest = specificProvider
    ? { [specificProvider]: AVAILABLE_PROVIDERS[specificProvider] }
    : AVAILABLE_PROVIDERS;

  // Run tests
  for (const [key, config] of Object.entries(providersToTest)) {
    const passed = await testProvider(
      config.class,
      config.name,
      config.isStreaming,
      config.requiresThinking
    );
    results.push([key, passed]);
  }

  // Print summary
  console.log(`\n${'='.repeat(80)}`);
  console.log('📊 Test Summary');
  console.log(`${'='.repeat(80)}\n`);

  for (const [name, passed] of results) {
    const status = passed ? '✅ PASS' : '❌ FAIL';
    console.log(`${status} - ${name}`);
  }

  const totalPassed = results.filter(([_, passed]) => passed).length;
  const totalTests = results.length;

  console.log(`\nTotal: ${totalPassed}/${totalTests} tests passed`);

  // Exit with appropriate code
  process.exit(totalPassed === totalTests ? 0 : 1);
}

main().catch(console.error);

File: python/scripts/test_weather_tool.py
#!/usr/bin/env python3
"""
Test script for weather tool functionality.
Tests that providers correctly call the weather API with lat/lon/location_name.

Usage:
    python test_weather_tool.py                    # Run all tests
    python test_weather_tool.py --provider anthropic   # Test specific provider
    python test_weather_tool.py --list             # List available providers
"""

import os
import sys
import argparse

# Add parent directory to path so we can import from providers
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from posthog import Posthog
from providers.anthropic import AnthropicProvider
from providers.anthropic_streaming import AnthropicStreamingProvider
from providers.gemini import GeminiProvider
from providers.gemini_streaming import GeminiStreamingProvider
from providers.openai import OpenAIProvider
from providers.openai_streaming import OpenAIStreamingProvider
from providers.openai_chat import OpenAIChatProvider
from providers.openai_chat_streaming import OpenAIChatStreamingProvider
from providers.langchain import LangChainProvider
from providers.litellm_provider import LiteLLMProvider
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Define available providers
AVAILABLE_PROVIDERS = {
    'anthropic': (AnthropicProvider, "Anthropic Provider", False),
    'anthropic-streaming': (AnthropicStreamingProvider, "Anthropic Streaming Provider", True),
    'gemini': (GeminiProvider, "Google Gemini Provider", False),
    'gemini-streaming': (GeminiStreamingProvider, "Google Gemini Streaming Provider", True),
    'openai': (OpenAIProvider, "OpenAI Responses API Provider", False),
    'openai-streaming': (OpenAIStreamingProvider, "OpenAI Responses API Streaming Provider", True),
    'openai-chat': (OpenAIChatProvider, "OpenAI Chat Completions Provider", False),
    'openai-chat-streaming': (OpenAIChatStreamingProvider, "OpenAI Chat Completions Streaming Provider", True),
    'langchain': (LangChainProvider, "LangChain Provider", False),
    'litellm': (LiteLLMProvider, "LiteLLM Provider", False),
}

def init_posthog():
    """Initialize PostHog client"""
    posthog_api_key = os.getenv("POSTHOG_API_KEY")
    posthog_host = os.getenv("POSTHOG_HOST", "https://us.i.posthog.com")

    if not posthog_api_key:
        print("⚠️  Warning: POSTHOG_API_KEY not set, analytics will not be tracked")
        return None

    return Posthog(
        project_api_key=posthog_api_key,
        host=posthog_host
    )

def test_provider(provider_class, provider_name, use_streaming=False):
    """Test a provider with weather tool calls"""
    print(f"\n{'='*80}")
    print(f"🧪 Testing {provider_name}")
    print(f"{'='*80}\n")

    # Initialize PostHog
    posthog_client = init_posthog()

    if not posthog_client:
        print("❌ Cannot run test without PostHog client")
        return False

    # Initialize provider
    provider = provider_class(posthog_client)

    # Test query
    test_query = "What's the weather in Dublin, Ireland?"
    print(f"📝 Query: {test_query}\n")

    try:
        if use_streaming:
            print("🔄 Streaming response:\n")
            full_response = ""
            for chunk in provider.chat_stream(test_query):
                print(chunk, end='', flush=True)
                full_response += chunk
            print("\n")
            response = full_response
        else:
            print("💬 Response:\n")
            response = provider.chat(test_query)
            print(response)
            print()

        # Check if response contains weather data
        if "°C" in response or "°F" in response:
            print("✅ Test PASSED - Weather data received!")

            # Check if location name is used
            if "Dublin" in response:
                print("✅ Location name displayed correctly!")
            else:
                print("⚠️  Warning: Location name might not be displayed")

            return True
        else:
            print("❌ Test FAILED - No weather data in response")
            return False

    except Exception as e:
        print(f"❌ Test FAILED with error: {str(e)}")
        import traceback
        traceback.print_exc()
        return False
    finally:
        # Shutdown PostHog
        if posthog_client:
            posthog_client.shutdown()

def list_providers():
    """List all available providers"""
    print("\n" + "="*80)
    print("📋 Available Providers")
    print("="*80 + "\n")

    for key, (_, name, is_streaming) in AVAILABLE_PROVIDERS.items():
        stream_indicator = " (streaming)" if is_streaming else ""
        print(f"  • {key:25} - {name}{stream_indicator}")

    print()

def main():
    """Run tests based on CLI arguments"""
    parser = argparse.ArgumentParser(
        description="Test weather tool functionality across different AI providers",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python test_weather_tool.py                          # Run all tests
  python test_weather_tool.py --provider anthropic     # Test Anthropic only
  python test_weather_tool.py --provider openai-chat   # Test OpenAI Chat only
  python test_weather_tool.py --list                   # List available providers
        """
    )

    parser.add_argument(
        '--provider', '-p',
        choices=list(AVAILABLE_PROVIDERS.keys()),
        help='Test a specific provider'
    )

    parser.add_argument(
        '--list', '-l',
        action='store_true',
        help='List available providers and exit'
    )

    args = parser.parse_args()

    # Handle --list flag
    if args.list:
        list_providers()
        sys.exit(0)

    print("\n" + "="*80)
    print("🚀 Weather Tool Test Suite")
    print("="*80)

    results = []

    # Determine which providers to test
    if args.provider:
        providers_to_test = {args.provider: AVAILABLE_PROVIDERS[args.provider]}
    else:
        providers_to_test = AVAILABLE_PROVIDERS

    # Run tests
    for key, (provider_class, provider_name, use_streaming) in providers_to_test.items():
        results.append((key, test_provider(provider_class, provider_name, use_streaming)))

    # Print summary
    print("\n" + "="*80)
    print("📊 Test Summary")
    print("="*80 + "\n")

    for name, passed in results:
        status = "✅ PASS" if passed else "❌ FAIL"
        print(f"{status} - {name}")

    total_passed = sum(1 for _, passed in results if passed)
    total_tests = len(results)

    print(f"\nTotal: {total_passed}/{total_tests} tests passed")

    # Exit with appropriate code
    sys.exit(0 if total_passed == total_tests else 1)

if __name__ == "__main__":
    main()
